core/src/trezor/ui/text.py

```diff
@@ -29,35 +31,88 @@ def render_text(
     offset_x: int = TEXT_MARGIN_LEFT,
     offset_y: int = TEXT_HEADER_HEIGHT + TEXT_LINE_HEIGHT,
     offset_x_max: int = ui.WIDTH,
+    pre_linebreaked: bool = False,
 ) -> None:
+    if not pre_linebreaked:
+        lines = break_lines(
+            words, new_lines, max_lines, font, fg, offset_x, offset_y, offset_x_max
+        )
+        words = [word for line in lines for word in line]
+
     # initial rendering state
     INITIAL_OFFSET_X = offset_x
     offset_y_max = TEXT_HEADER_HEIGHT + (TEXT_LINE_HEIGHT * max_lines)
 
-    FONTS = (ui.NORMAL, ui.BOLD, ui.MONO)
+    for word in words:
+        if isinstance(word, int):
+            if word is BR or word is BR_HALF:
+                # line break or half-line break
+                if offset_y > offset_y_max:
+                    return
+                offset_x = INITIAL_OFFSET_X
+                offset_y += TEXT_LINE_HEIGHT if word is BR else TEXT_LINE_HEIGHT_HALF
+            elif word in _FONTS:
+                # change of font style
+                font = word
+            else:
+                # change of foreground color
+                fg = word
+            continue
+
+        ui.display.text(offset_x, offset_y, word, font, fg, bg)
+        offset_x += ui.display.text_width(word, font)
+
+
+def break_lines(
+    words: Sequence[TextContent],
+    new_lines: bool,
+    max_lines: Optional[int],
+    font: int = ui.NORMAL,
+    fg: int = ui.FG,
+    offset_x: int = TEXT_MARGIN_LEFT,
+    offset_y: int = TEXT_HEADER_HEIGHT + TEXT_LINE_HEIGHT,
+    offset_x_max: int = ui.WIDTH,
+) -> List[List[TextContent]]:
```

Comment:
why return List[List] if you convert it to a flat list anyway?

Reply:
In `apps/common/signverify.py` (and then `apps/common/layout`) we first call `chunks` on the result before passing it to `render_text` which flattens it.